lib/system/freertos/xlnx_common/irq.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2016 - 2017, Xilinx Inc. and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+/*
+ * @file	generic/xlnx_common/irq.c
+ * @brief	generic libmetal Xilinx irq controller definitions.
+ */
+
+#include <errno.h>
+#include <metal/irq.h>
+#include <metal/sys.h>
+#include <metal/log.h>
+#include <metal/mutex.h>
+#include <metal/list.h>
+#include <metal/utilities.h>
+#include <metal/alloc.h>
+
+#define MAX_IRQS XLNX_MAXIRQS
+
+static struct metal_irq irqs[MAX_IRQS]; /**< Linux IRQs array */
+
+static void metal_xlnx_irq_set_enable(struct metal_irq_controller *irq_cntr,
+				     int irq, unsigned int state)
+{
+	if (irq < irq_cntr->irq_base ||
+	    irq > irq_cntr->irq_base + irq_cntr->irq_num) {
+		metal_log(METAL_LOG_ERROR, "%s: invalid irq %d\n",
+			  __func__, irq);
+		return;
+	} else if (state == METAL_IRQ_ENABLE) {
+		sys_irq_enable((unsigned int)irq);
+	} else {
+		sys_irq_disable((unsigned int)irq);
+	}
+}
+
+/**< Xilinx common platform IRQ controller */
+static METAL_IRQ_CONTROLLER_DECLARE(xlnx_irq_cntr,
+				    0, MAX_IRQS,
+				    NULL, metal_xlnx_irq_set_enable,
+				    irqs)
+
+/**
+ * @brief default handler
+ */
+void metal_xlnx_irq_isr(unsigned int vector)
+{
+	struct metal_irq *irq_data;
+
+	if (vector > MAX_IRQS) {
+		return;
+	}
+	irq_data = &irqs[vector];
+	if (irq_data->hd) {
+		(void)irq_data->hd(vector, irq_data->arg);
+	}
```

Comment:
reuse metal_irq_handle here

Reply:
I don't really get this comment, do  you mean prefer to use a metal_irq_handler to get the hd first?
Will the compiler do the optimisation automatically?

Reply:
we can simplify to one line code:
metal_irq_handle(irqs[vector], vector);